### PR TITLE
fix: checking whether interaction is in progress inside interceptor

### LINKED ISSUE #5146: Interaction is currently in progress
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.28.1

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

2.4.1

### Public or Confidential Client?

Public

### Description

Incognito mode in Chrome gives out interraction in progress error. User has conditional access policies to prevent the 24 hours logout but without luck.

image

In the custom guards and directives I check for the InteractionStatus.None

image

image

### Error Message

```javascript
ERROR Error: Uncaught (in promise): BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: aka.ms/msaljs/browser-errors.
BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: aka.ms/msaljs/browser-errors.
    at BrowserAuthError.AuthError [as constructor] (vendor.js:19617:20)
    at new BrowserAuthError (vendor.js:6921:24)
    at BrowserAuthError.createInteractionInProgressError (vendor.js:7009:12)
    at BrowserCacheManager.setInteractionInProgress (vendor.js:4601:89)
    at ClientApplication.preflightInteractiveRequest (vendor.js:2212:27)
    at ClientApplication.preflightBrowserEnvironmentCheck (vendor.js:2195:12)
    at PublicClientApplication.<anonymous> (vendor.js:1557:14)
    at step (vendor.js:1203:17)
    at Object.next (vendor.js:1134:14)
    at vendor.js:1106:67
    at resolvePromise (VM27:9098:19)
    at new ZoneAwarePromise (VM27:9360:9)
    at __awaiter (VM28:formatted:20:NaN)
    at ClientApplication.acquireTokenRedirect (VM28:formatted:20:NaN)
    at MsalService.acquireTokenRedirect (VM28:formatted:20:NaN)
    at MsalInterceptor.acquireTokenInteractively (VM28:formatted:20:NaN)
    at VM28:formatted:20:NaN
    at VM28:formatted:20:NaN
    at OperatorSubscriber._error (VM28:formatted:20:NaN)
    at OperatorSubscriber.error (VM28:formatted:20:NaN)
handleError @ VM28:formatted:20
next @ VM28:formatted:20
next @ VM28:formatted:20
_next @ VM28:formatted:20
next @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
errorContext @ VM28:formatted:20
next @ VM28:formatted:20
emit @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
invoke @ VM27:8086
run @ VM27:7827
runOutsideAngular @ VM28:formatted:20
onHandleError @ VM28:formatted:20
handleError @ VM27:8090
runGuarded @ VM27:7843
api.microtaskDrainDone @ VM27:8935
drainMicroTaskQueue @ VM27:8335
invokeTask @ VM27:8214
ZoneTask.invoke @ VM27:8197
data.args.<computed> @ VM27:10505
setInterval (async)
scheduleTask @ VM27:10508
scheduleTask @ VM27:8107
onScheduleTask @ VM27:8007
scheduleTask @ VM27:8102
scheduleTask @ VM27:7928
scheduleMacroTask @ VM27:7957
scheduleMacroTaskWithCurrentZone @ VM27:8460
(anonymous) @ VM27:10550
proto.<computed> @ VM27:8813
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
SilentHandler.monitorIframeForHash @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
fulfilled @ VM28:formatted:20
invoke @ VM27:8086
onInvoke @ VM28:formatted:20
invoke @ VM27:8086
run @ VM27:7827
(anonymous) @ VM27:9171
invokeTask @ VM27:8119
onInvokeTask @ VM28:formatted:20
invokeTask @ VM27:8119
runTask @ VM27:7880
drainMicroTaskQueue @ VM27:8328
Promise.then (async)
nativeScheduleMicroTask @ VM27:8299
scheduleMicroTask @ VM27:8310
scheduleTask @ VM27:8109
onScheduleTask @ VM27:8007
scheduleTask @ VM27:8102
scheduleTask @ VM27:7928
scheduleMicroTask @ VM27:7953
scheduleResolveOrReject @ VM27:9159
resolvePromise @ VM27:9087
(anonymous) @ VM27:8993
(anonymous) @ VM27:9010
Promise.then (async)
(anonymous) @ VM27:9457
ZoneAwarePromise @ VM27:9358
Ctor.then @ VM27:9456
resolvePromise @ VM27:9048
(anonymous) @ VM27:8993
(anonymous) @ VM27:9010
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
BrowserCrypto.getSubtleCryptoDigest @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
BrowserCrypto.sha256Digest @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
PkceGenerator.generateCodeChallengeFromVerifier @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
PkceGenerator.generateCodes @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
CryptoOps.generatePkceCodes @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
StandardInteractionClient.initializeAuthorizationCodeRequest @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
SilentIframeClient.silentTokenHelper @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
fulfilled @ VM28:formatted:20
invoke @ VM27:8086
onInvoke @ VM28:formatted:20
invoke @ VM27:8086
run @ VM27:7827
(anonymous) @ VM27:9171
invokeTask @ VM27:8119
onInvokeTask @ VM28:formatted:20
invokeTask @ VM27:8119
runTask @ VM27:7880
drainMicroTaskQueue @ VM27:8328
Promise.then (async)
nativeScheduleMicroTask @ VM27:8299
scheduleMicroTask @ VM27:8310
scheduleTask @ VM27:8109
onScheduleTask @ VM27:8007
scheduleTask @ VM27:8102
scheduleTask @ VM27:7928
scheduleMicroTask @ VM27:7953
scheduleResolveOrReject @ VM27:9159
resolvePromise @ VM27:9087
(anonymous) @ VM27:8993
(anonymous) @ VM27:9010
Promise.then (async)
(anonymous) @ VM27:9457
ZoneAwarePromise @ VM27:9358
Ctor.then @ VM27:9456
resolvePromise @ VM27:9048
(anonymous) @ VM27:8993
(anonymous) @ VM27:9010
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
adopt @ VM28:formatted:20
step @ VM28:formatted:20
fulfilled @ VM28:formatted:20
invoke @ VM27:8086
onInvoke @ VM28:formatted:20
invoke @ VM27:8086
run @ VM27:7827
(anonymous) @ VM27:9171
invokeTask @ VM27:8119
onInvokeTask @ VM28:formatted:20
invokeTask @ VM27:8119
runTask @ VM27:7880
drainMicroTaskQueue @ VM27:8328
Promise.then (async)
nativeScheduleMicroTask @ VM27:8299
scheduleMicroTask @ VM27:8310
scheduleTask @ VM27:8109
onScheduleTask @ VM27:8007
scheduleTask @ VM27:8102
scheduleTask @ VM27:7928
scheduleMicroTask @ VM27:7953
scheduleResolveOrReject @ VM27:9159
resolvePromise @ VM27:9087
(anonymous) @ VM27:8993
(anonymous) @ VM27:9010
Promise.then (async)
(anonymous) @ VM27:9457
ZoneAwarePromise @ VM27:9358
Ctor.then @ VM27:9456
resolvePromise @ VM27:9048
(anonymous) @ VM27:8993
(anonymous) @ VM27:9010
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
adopt @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
FetchClient.sendPostRequestAsync @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
NetworkManager.sendPostRequest @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
ZoneAwarePromise @ VM27:9358
__awaiter @ VM28:formatted:20
BaseClient.executePostToTokenEndpoint @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
step @ VM28:formatted:20
(anonymous) @ VM28:formatted:20
fulfilled @ VM28:formatted:20
invoke @ VM27:8086
onInvoke @ VM28:formatted:20
invoke @ VM27:8086
run @ VM27:7827
(anonymous) @ VM27:9171
invokeTask @ VM27:8119
onInvokeTask @ VM28:formatted:20
invokeTask @ VM27:8119
runTask @ VM27:7880
drainMicroTaskQueue @ VM27:8328
invokeTask @ VM27:8214
ZoneTask.invoke @ VM27:8197
data.args.<computed> @ VM27:10505
Show 112 more frames
20:32:54.063 
```

_No response_

### Msal Logs

_No response_

### MSAL Configuration

```javascript
export function MSALInstanceFactory(): IPublicClientApplication {
  return new PublicClientApplication({
    auth: {
      clientId: 'xxxx',
      authority: 'https://login.microsoftonline.com/xxxxx',
      redirectUri: environment.msalRedirectUri,
      postLogoutRedirectUri: '/'
    },
    cache: {
      cacheLocation: BrowserCacheLocation.LocalStorage,
      storeAuthStateInCookie: isIE // set to true for IE 11. Remove this line to use Angular Universal
    }
    // system: {
    //   loggerOptions: {
    //     loggerCallback,
    //     logLevel: LogLevel.Info,
    //     piiLoggingEnabled: false
    //   }
    // }
  });
}

export function MSALInterceptorConfigFactory(): MsalInterceptorConfiguration {
  const protectedResourceMap = new Map<string, Array<string>>();

  protectedResourceMap.set('https://graph.microsoft.com/v1.0/me', ['user.read']);
  protectedResourceMap.set(xxxxx, ['xxxxx']);

  return {
    interactionType: InteractionType.Redirect,
    protectedResourceMap
  };
}

export function MSALGuardConfigFactory(): MsalGuardConfiguration {
  return {
    interactionType: InteractionType.Redirect,
    authRequest: {
      scopes: ['user.read']
    },
    loginFailedRoute: '/'
  };
}
```

### Relevant Code Snippets

```javascript
@NgModule({
  declarations: [UserRolesDirective],
  imports: [CommonModule],
  providers: [AuthGuard, RolesGuard],
  exports: [UserRolesDirective]
})
export class AuthModule {
  static forRoot(): ModuleWithProviders<AuthModule> {
    return {
      ngModule: AuthModule,
      providers: [
        {
          provide: HTTP_INTERCEPTORS,
          useClass: MsalInterceptor,
          multi: true
        },
        {
          provide: MSAL_INSTANCE,
          useValue: MSALInstanceFactory()
        },
        {
          provide: MSAL_GUARD_CONFIG,
          useValue: MSALGuardConfigFactory()
        },
        {
          provide: MSAL_INTERCEPTOR_CONFIG,
          useValue: MSALInterceptorConfigFactory()
        },
        MsalService,
        MsalGuard,
        MsalBroadcastService
      ]
    };
  }
}
```

### Reproduction Steps

Using standard @azure/msal-angular configuration and wait for the refresh token to expire (24h)

### Expected Behavior

Shoul not throw "interaction_in_progress"

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome

### Regression

_No response_

### Source

External (Customer)

### PR DESCRIPTION
**Pull request by yksht**
PR for #5146 

Lets try to get progress here :) Im also facing the issue on my project

I saw comments in previous PR and moved interaction in progress check before invoking acquireTokenSilent.
But I didn't get idea from the comment:
> start listening for inProgress events in the constuctor, and main an internal field (i.e. this.currentInteractionStatus), instead of listening only when acquireTokenInteractively is called

I don't see performance impact here

### CODE REVIEW

File: lib/msal-angular/src/msal.interceptor.ts [hunk 1]
```diff
@@ -65,26 +67,33 @@ export class MsalInterceptor implements HttpInterceptor {
...
-                    this.authService.getLogger().verbose("Interceptor - setting authorization headers");
-                    const headers = req.headers
-                        .set("Authorization", `Bearer ${result.accessToken}`);
-
-                    const requestClone = req.clone({headers});
-                    return next.handle(requestClone);
+                filter((status: InteractionStatus) => status === InteractionStatus.None),
+                take(1),
+                switchMap(() => {
+                    return this.authService.acquireTokenSilent({...authRequest, scopes, account })
```
Comment (tnorling) [inline lib/msal-angular/src/msal.interceptor.ts]: The silent call is allowed when interaction is in progress we should continue to let it execute

File: lib/msal-angular/src/msal.interceptor.ts [hunk 2]
```diff
@@ -65,26 +67,33 @@ export class MsalInterceptor implements HttpInterceptor {
...
+                        .pipe(
+                            catchError(() => {
+                                this.authService.getLogger().error("Interceptor - acquireTokenSilent rejected with error. Invoking interaction to resolve.");
+                                return this.acquireTokenInteractively(authRequest, scopes);
```
Comment (tnorling) [inline lib/msal-angular/src/msal.interceptor.ts]: What happens if interaction was not in progress when `acquireTokenSilent` was invoked but then it is in progress when it reaches here?

Ideally I'd like to see the flow look something like this:
1. Invoke `acquireTokenSilent` regardless of interaction status
2. When ATS throws, check interaction status
3. If interaction is not currently in progress invoke interaction
4. If interaction is in progress, wait until it is done then go back to step 1

**Review by tnorling — COMMENTED** [review]
Thanks for the PR! A couple comments regarding the approach here

File: lib/msal-angular/src/msal.interceptor.ts (hunk 2 above)
Comment (yksht) [reply lib/msal-angular/src/msal.interceptor.ts]: @tnorling made some changes. Now it follows your logic

Comment (tnorling) [reply lib/msal-angular/src/msal.interceptor.ts]: Looks good, thanks! Any tests you can add?

**Comment by yksht** [comment]
@tnorling tests added

**Review by tnorling — APPROVED** [review]
Looks good to me! @jo-arroyo Could you please take a look at this as well?

**Comment by tnorling** [comment]
@yksht Please also run `npm run beachball:change` from the root of the repo and follow the prompts to generate the changefile when you get a chance. Thanks!

**Comment by yksht** [comment]
@tnorling All done. Thanks for help :)

**Review by derisen — APPROVED** [review]

**Review by jo-arroyo — APPROVED** [review]
Looks good to me!

**Comment by loebi-ch** [comment]
This fix seems to break the bootstraping of my Angular application.

With version 2.5.3 of @azure/msal-angular all works, with version 2.5.4 of @azure/msal-angular with this fix, the browser gets stuck on bootstraping. It's loading until a "Out of Memory" exception is shown in the browser...

I tried to get some information in Developer tools, but the browser is kind of freezed (stack overflowed).

This is how I use the MSAL interceptor in my Angular application:

```
providers: [
        {
            provide: MSAL_INSTANCE,
            useFactory: _ => {
                return new PublicClientApplication({
                    auth: {
                        clientId: '********-****-****-****-************',
                        authority: 'https://login.microsoftonline.com/organizations',
                        redirectUri: location.protocol + '//' + location.host,
                        navigateToLoginRequestUrl: false
                    },
                    cache: {
                        cacheLocation: BrowserCacheLocation.SessionStorage,
                        storeAuthStateInCookie: window.navigator.userAgent.indexOf("MSIE ") > -1 || window.navigator.userAgent.indexOf("Trident/") > -1
                    }
                });
            }
        },
        {
            provide: MSAL_INTERCEPTOR_CONFIG, useValue: {
                interactionType: InteractionType.Redirect,
                protectedResourceMap: new Map([
                    //... my protected resources
                ])
            } as MsalInterceptorConfiguration
        },
        {
            provide: HTTP_INTERCEPTORS,
            useClass: MsalInterceptor,
            multi: true
        }
    ]
```

Are others facing the same issues with version 2.5.4 of @azure/msal-angular?
Or do I have to change some configuration of the MSAL interceptor?

**Comment by yksht** [comment]
@loebi-ch 
This fix affects only `intercept` method of MsalInterceptor so it should not take effect on instanciating MsalInterceptor.
I have updated my application with version 2.5.4 and have not such problems even though in general have same providers in the app. 

The problem can be if interceptor cannot acquire access token for any reason. So you need to check requests to your authority. If you cannot use devtools then go with a tool like Fiddler. I think you will find a problem when investigate these requests.

@tnorling 
In any case there is definitely a problem if token acquiring always fails - we get infinit loop inside `intercept` method. 
What do you think? I can create a new task about implementing some sort of retry strategy for this place.

**Comment by loebi-ch** [comment]
@yksht Token acquiring definitely fails because the user is not logged in yet...

**Comment by yksht** [comment]
@loebi-ch
In case interceptor can not acquire token silently (your case) it will do this interactively and according to your settings (`InteractionType.Redirect`) should redirect you to authority login page if you have no active session.
But as I understand you do not have redirects at all. You have "Out of Memory" exception instead. So for me it looks like interceptor is not a reason of your problem.

Eventualy there is one usecase when this can happen, but both conditions should met:
1. InteractionStatus stucks and never became 'None' - ensure you properly bootstraped MsalRedirectComponent (or manualy call handleRedirectObservable method of MsalService before any requests)
2. You have corrupted session data and library thinks you have active session but you are actually not - try to remove all app's cookies and cookies for login.microsoftonline.com domain

I want to figure out if changes from this PR can trigger such behavior, but probably this is not best place to have the chat :) please contact me in linkedin (you can find link in my profile) and let's continue there

**Pull request merged**